<a href="https://colab.research.google.com/github/iwishkim/stock_market_analysis/blob/main/pandas_exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/zzhining/ml_basic/main/dataset/insurance.csv')

In [4]:
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [5]:
# 주요 통계지표
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [6]:
df['double bmi'] = df['bmi'] * 2

In [7]:
df.head()

age     sex     bmi  children smoker     region      charges  double bmi
0   19  female  27.900         0    yes  southwest  16884.92400       55.80
1   18    male  33.770         1     no  southeast   1725.55230       67.54
2   28    male  33.000         3     no  southeast   4449.46200       66.00
3   33    male  22.705         0     no  northwest  21984.47061       45.41
4   32    male  28.880         0     no  northwest   3866.85520       57.76

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [8]:
df['debt'] = 0
df.head()

age     sex     bmi  children smoker     region      charges  double bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     0  
2     0  
3     0  
4     0

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [9]:
df['debt'] = np.arange(len(df))
df.head()

age     sex     bmi  children smoker     region      charges  double bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     1  
2     2  
3     3  
4     4

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [10]:
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double bmi        67.54
debt                  1
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [11]:
# axis = 1: 축을 컬럼방향으로 설정하기 위한 매개변수
# inplace = True: df 객체에 있는 값을 연산 결과로 대체
df.drop(['double bmi'], axis=1, inplace = True)

In [12]:
df.head()

age     sex     bmi  children smoker     region      charges  debt
0   19  female  27.900         0    yes  southwest  16884.92400     0
1   18    male  33.770         1     no  southeast   1725.55230     1
2   28    male  33.000         3     no  southeast   4449.46200     2
3   33    male  22.705         0     no  northwest  21984.47061     3
4   32    male  28.880         0     no  northwest   3866.85520     4

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [13]:
df[['age', 'sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [14]:
df['age'] = df['age'] + 1 

In [15]:
df.head()

age     sex     bmi  children smoker     region      charges  debt
0   20  female  27.900         0    yes  southwest  16884.92400     0
1   19    male  33.770         1     no  southeast   1725.55230     1
2   29    male  33.000         3     no  southeast   4449.46200     2
3   34    male  22.705         0     no  northwest  21984.47061     3
4   33    male  28.880         0     no  northwest   3866.85520     4

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [16]:
# round(): 반올림
df['charges'].round(2)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: float64

In [17]:
test_num = 123.45678

def sosu(data):
    return format(data, ".2f")

sosu(5883829.2326)

'5883829.23'

In [18]:
df['charges'] = df['charges'].apply(sosu)

In [19]:
df.head()

age     sex     bmi  children smoker     region   charges  debt
0   20  female  27.900         0    yes  southwest  16884.92     0
1   19    male  33.770         1     no  southeast   1725.55     1
2   29    male  33.000         3     no  southeast   4449.46     2
3   34    male  22.705         0     no  northwest  21984.47     3
4   33    male  28.880         0     no  northwest   3866.86     4

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [20]:
df.sort_index(axis = 1, inplace=True)

In [21]:
df.head()

age     bmi   charges  children  debt     region     sex smoker
0   20  27.900  16884.92         0     0  southwest  female    yes
1   19  33.770   1725.55         1     1  southeast    male     no
2   29  33.000   4449.46         3     2  southeast    male     no
3   34  22.705  21984.47         0     3  northwest    male     no
4   33  28.880   3866.86         0     4  northwest    male     no

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [22]:
df.sort_values(by ='age')

age     bmi   charges  children  debt     region     sex smoker
1248   19  39.820   1633.96         0  1248  southeast  female     no
482    19  31.350   1622.19         0   482  southeast  female     no
492    19  25.080   2196.47         0   492  northeast  female     no
525    19  33.880  11482.63         0   525  southeast  female     no
529    19  25.460   1708.00         0   529  northeast    male     no
...   ...     ...       ...       ...   ...        ...     ...    ...
398    65  25.600  14988.43         2   398  southwest    male     no
335    65  34.500  13822.80         0   335  southwest    male     no
378    65  30.115  16455.71         3   378  northwest  female     no
1265   65  23.760  26926.51         0  1265  southeast    male    yes
635    65  38.190  14410.93         0   635  northeast    male     no

[1338 rows x 8 columns]

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [23]:
df['age'].mean()

40.20702541106129

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


In [33]:
df['smoker'].value_counts()

no     1064
yes     274
Name: smoker, dtype: int64

In [26]:
df['sex'].value_counts()

male      676
female    662
Name: sex, dtype: int64

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

In [29]:
df[df['children'] ==3]

age     bmi   charges  children  debt     region     sex smoker
2      29  33.000   4449.46         3     2  southeast    male     no
7      38  27.740   7281.51         3     7  northwest  female     no
25     60  27.720  14001.13         3    25  southeast  female     no
36     63  32.965  15612.19         3    36  northwest  female     no
54     41  28.690   8059.68         3    54  northwest  female     no
...   ...     ...       ...       ...   ...        ...     ...    ...
1301   63  30.875  46718.16         3  1301  northwest    male    yes
1314   31  23.655  18765.88         3  1314  northwest  female    yes
1320   32  31.065   5425.02         3  1320  northwest    male     no
1332   53  44.700  11411.68         3  1332  southwest  female     no
1333   51  30.970  10600.55         3  1333  northwest    male     no

[157 rows x 8 columns]

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

In [30]:
df[df['children'] >=5]

age     bmi   charges  children  debt     region     sex smoker
32     20  28.600   4687.80         5    32  southwest  female     no
71     32  28.500   6799.46         5    71  northeast    male     no
166    21  37.000   4830.63         5   166  southwest  female     no
413    26  23.900   5080.10         5   413  southwest    male     no
425    46  24.310   9788.87         5   425  southeast    male     no
438    53  46.750  12592.53         5   438  southeast  female     no
568    50  31.900  11552.90         5   568  southwest  female     no
640    34  42.400   6666.24         5   640  southwest    male     no
877    34  33.440   6653.79         5   877  southeast    male     no
932    47  25.800  10096.97         5   932  southwest    male     no
937    40  24.225   8965.80         5   937  northwest  female     no
969    40  34.320   8596.83         5   969  southeast  female     no
984    21  30.115   4915.06         5   984  northeast    male     no
1085   40  18.300  19023.26         5  1085  southwest  female    yes
1116   42  29.640   9222.40         5  1116  northeast    male     no
1130   40  23.870   8582.30         5  1130  southeast  female     no
1245   29  24.300   5615.37         5  1245  southwest    male     no
1272   44  25.520  14478.33         5  1272  southeast    male     no

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

In [32]:
df[df['bmi'] <17]

age     bmi  charges  children  debt     region     sex smoker
172    19  15.960  1694.80         0   172  northeast    male     no
428    22  16.815  3167.46         1   428  northeast  female     no
1226   39  16.815  6640.54         2  1226  northeast    male     no

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

In [41]:
df['charges'] = df['charges'].astype(float)

In [42]:
df.groupby('sex')

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   bmi       1338 non-null   float64
 2   charges   1338 non-null   float64
 3   children  1338 non-null   int64  
 4   debt      1338 non-null   int64  
 5   region    1338 non-null   object 
 6   sex       1338 non-null   object 
 7   smoker    1338 non-null   object 
dtypes: float64(2), int64(3), object(3)
memory usage: 83.8+ KB


In [44]:
df.groupby('sex')['charges'].mean()

sex
female    12569.578731
male      13956.751317
Name: charges, dtype: float64

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

In [45]:
df.groupby('smoker')['charges'].mean()

smoker
no      8434.268299
yes    32050.231898
Name: charges, dtype: float64

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

In [46]:
df.groupby('region')['charges'].mean()

region
northeast    13406.384599
northwest    12417.575169
southeast    14735.411484
southwest    12346.937508
Name: charges, dtype: float64

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

In [47]:
df.groupby(['smoker','region'],)['charges'].mean()

smoker  region   
no      northeast     9165.531829
        northwest     8556.463558
        southeast     8032.216264
        southwest     8019.284569
yes     northeast    29673.536269
        northwest    30192.002759
        southeast    34844.997143
        southwest    32269.063966
Name: charges, dtype: float64

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

In [48]:
df.groupby(['smoker','region'],)['charges'].mean()

smoker  region   
no      northeast     9165.531829
        northwest     8556.463558
        southeast     8032.216264
        southwest     8019.284569
yes     northeast    29673.536269
        northwest    30192.002759
        southeast    34844.997143
        southwest    32269.063966
Name: charges, dtype: float64

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

In [51]:
df.groupby(['smoker','region']).size()

smoker  region   
no      northeast    257
        northwest    267
        southeast    273
        southwest    267
yes     northeast     67
        northwest     58
        southeast     91
        southwest     58
dtype: int64

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>

In [56]:
smgu = lambda sm : ('비흡연자') if(sm=='no') else ('흡연자')
df['group_val']=df['smoker'].apply(smgu)

In [57]:
df['group_val']=df['group_val'].apply(lambda bool: 'smoker' if bool=='yes' else 'non-smoker')
df.head(3)

age    bmi   charges  children  debt     region     sex smoker   group_val
0   20  27.90  16884.92         0     0  southwest  female    yes  non-smoker
1   19  33.77   1725.55         1     1  southeast    male     no  non-smoker
2   29  33.00   4449.46         3     2  southeast    male     no  non-smoker

In [58]:
df['group_val'] = df['group_val'].apply(lambda bool: 'smoker' if bool=='yes' else 'non-smoker')
df.head(3)

age    bmi   charges  children  debt     region     sex smoker   group_val
0   20  27.90  16884.92         0     0  southwest  female    yes  non-smoker
1   19  33.77   1725.55         1     1  southeast    male     no  non-smoker
2   29  33.00   4449.46         3     2  southeast    male     no  non-smoker